src/components/views/rooms/Autocomplete.js

```diff
@@ -42,38 +54,67 @@ export default class Autocomplete extends React.Component {
         });
     }
 
-    onUpArrow() {
-        this.setState({selectionOffset: this.state.selectionOffset - 1});
+    countCompletions(): number {
+        return this.state.completions.map(completionResult => {
+            return completionResult.completions.length;
+        }).reduce((l, r) => l + r);
+    }
+
+    // called from MessageComposerInput
+    onUpArrow(): boolean {
+        let completionCount = this.countCompletions(),
+            selectionOffset = (completionCount + this.state.selectionOffset - 1) % completionCount;
+        this.setSelection(selectionOffset);
         return true;
     }
 
-    onDownArrow() {
-        this.setState({selectionOffset: this.state.selectionOffset + 1});
+    // called from MessageComposerInput
+    onDownArrow(): boolean {
+        let completionCount = this.countCompletions(),
+            selectionOffset = (this.state.selectionOffset + 1) % completionCount;
+        this.setSelection(selectionOffset);
         return true;
     }
 
+    /** called from MessageComposerInput
+     * @returns {boolean} whether confirmation was handled
+     */
+    onConfirm(): boolean {
+        if (this.countCompletions() === 0)
+            return false;
+
+        let selectedCompletion = this.state.completionList[this.state.selectionOffset];
+        this.props.onConfirm(selectedCompletion.range, selectedCompletion.completion);
+
+        return true;
+    }
+
+    setSelection(selectionOffset: number) {
+        this.setState({selectionOffset});
+    }
+
     render() {
         let position = 0;
         let renderedCompletions = this.state.completions.map((completionResult, i) => {
             let completions = completionResult.completions.map((completion, i) => {
-                let Component = completion.component;
                 let className = classNames('mx_Autocomplete_Completion', {
-                    'selected': position == this.state.selectionOffset
+                    'selected': position === this.state.selectionOffset,
                 });
                 let componentPosition = position;
                 position++;
-                if(Component) {
-                    return Component;
-                }
-                
+
+                let onMouseOver = () => this.setSelection(componentPosition),
+                    onClick = () => {
+                        this.setSelection(componentPosition);
+                        this.onConfirm();
+                    };
```

Comment:
This is just a style point, but what's the advantage of putting both declarations in the same statement? To me this looks confusing because it make the onMouseOver declaration looks like it's related to the onClick function.